stix2/registration.py

```diff
@@ -1,11 +1,72 @@
-import itertools
 import re
 
 from . import registry, version
 from .base import _DomainObject
 from .exceptions import DuplicateRegistrationError
-from .properties import _validate_type
-from .utils import PREFIX_21_REGEX, get_class_hierarchy_names
+from .properties import (
+    ListProperty, ObjectReferenceProperty, ReferenceProperty, _validate_type,
+)
+from .utils import PREFIX_21_REGEX
+
+
+def _validate_ref_props(props_map, version, is_observable20=False):
+    """
+    Validate that reference properties contain an expected type.
+
+    Properties ending in "_ref/s" need to be instances of specific types to
+    meet the specification requirements. For 2.0 and 2.1 conformance, these
+    properties are expected to be implemented with `ReferenceProperty` (or a
+    subclass thereof), except for the special case of STIX 2.0 observables
+    which must be implemented with `ObjectReferenceProperty`.
+
+    Args:
+        props_map (mapping): A mapping of STIX object properties to be checked.
+        version (str): Which STIX2 version the properties must conform to.
+        is_observable20 (bool): Flag for the STIX 2.0 observables special case.
+
+    Raises:
+        ValueError: If the properties do not conform.
+    """
+    if is_observable20:
+        ref_prop_type = ObjectReferenceProperty
+    else:
+        ref_prop_type = ReferenceProperty
+    for prop_name, prop_obj in props_map.items():
+        tail = prop_name.rsplit("_", 1)[-1]
+        if tail == "ref" and not isinstance(prop_obj, ref_prop_type):
+            raise ValueError(
+                f"{prop_name!r} is named like a reference property but is not "
+                f"a subclass of {ref_prop_type.__name__!r}.",
+            )
+        elif tail == "refs" and not all((
+            isinstance(prop_obj, ListProperty),
+            isinstance(getattr(prop_obj, "contained", None), ref_prop_type),
+        )):
```

Comment:
This use of `all()` undermines the normal short-circuiting behavior of boolean evaluation.  Here, all args must always be evaluated before `all()` itself can be called.  But you could take advantage of short-circuiting here to drop the `getattr()` and simplify the expression.  E.g.:

```python
        elif tail == "refs" and not (
            isinstance(prop_obj, ListProperty)
            and isinstance(prop_obj.contained, ref_prop_type)
        ):
```

Here, `prop_obj.contained` will not be evaluated unless the previous conjunct (`isinstance(prop_obj, ListProperty)`) evaluated to True.  Therefore, at the point of the attribute access, prop_obj must be a ListProperty and the access is guaranteed to work.  So you don't need `getattr()`.

Reply:
Done in the new diff